readme.md
# Leaflet.GlobeMiniMap
<a href="https://www.npmjs.com/package/leaflet-globe-minimap">
<img src="https://img.shields.io/npm/v/leaflet-globe-minimap.svg" alt="npm version">
</a>

[![GitHub last commit](https://img.shields.io/github/last-commit/syonfox/leaflet-globe-minimap)](https://github.com/syonfox/leaflet-globe-minimap/commits)

[//]: # ([![example workflow]&#40;https://github.com/syonfox/leaflet-globe-minimap/actions/workflows/node.js.yml/badge.svg&#41;]&#40;https://github.com/syonfox/leaflet-globe-minimap/actions&#41;)
[![github](https://img.shields.io/badge/GitHub-100000?style=for-the-badge&logo=github&logoColor=white)](https://github.com/syonfox/leaflet-globe-minimap)


Leaflet.GlobeMiniMap is a simple minimap control that places a 3d Globe in the corner of your map, centered on the same location as the main map.

![cursor_and_3d_minimap_and_minimap_ _bash_ _80x24_and_index_html_ _globe-minimap](https://cloud.githubusercontent.com/assets/1833820/10415088/cb1d45fe-6fb8-11e5-9903-2c2ec16fbabd.png)

[Try the example out out now on github pages](http://syonfox.github.io/leaflet-globe-minimap/example/)


## Using the GlobeMiniMap control

Leaflet.GlobeMiniMap requires d3.js & topojson.js.  You can load them from a CDN like this:

```
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/leaflet/1.9.3/leaflet.css" integrity="sha512-mD70nAW2ThLsWH0zif8JPbfraZ8hbCtjQ+5RU1m4+ztZq6/MymyZeB55pWsi4YAX+73yvcaJyk61mzfYMvtm9w==" crossorigin="anonymous" referrerpolicy="no-referrer" />
  <script src="https://cdnjs.cloudflare.com/ajax/libs/d3/3.5.5/d3.min.js"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/topojson/1.6.19/topojson.min.js"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/leaflet/1.9.3/leaflet.js" integrity="sha512-Dqm3h1Y4qiHUjbhxTuBGQsza0Tfppn53SHlu/uj1f+RT+xfShfe7r6czRf5r2NmllO2aKx+tYJgoxboOkn1Scg==" crossorigin="anonymous" referrerpolicy="no-referrer"></script>
```
Node doesnt need leaflet 1.9 just works with it.

The library is also looking for `/data/world.json` in order to load the world topojson file. You can use the option `topojsonSrc` to define another source.

Add the globe minimap with one line of code:

```
    var miniMap = new L.Control.GlobeMiniMap(options).addTo(map);
```

You can pass in an options object to define colors for the water, land and marker on the globe minimap:

```
{
  land:'#FFFF00',
  water:'#3333FF',
  marker:'#000000',
  topojsonSrc: 'path/to/world.json',
  durations: 1250,           // use to control how snapy the animation is
  width: 82,                 // the size, as of now they must be same as height
  height: 82,                // maybe we should renam radius and have one options
  onAdd: null,               // calback functions for listners
  onRemove: null,
}
```

The minimap is reactive, it only changes when the main map's center point changes.  You can't interact with the mini map to move the main map.


# Advanced Styling

I have added the option to use a radial gradiant to style the globeStroke.

This is what the gradient

```
 globeStroke: [
    [0, 'red'],
    [0.5, 'blue'],
    [1, 'green']
  ],
  drawGlobeGradient: true,
```
looks like see: the docs form more info https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/createRadialGradient#examples
each array adds a colorStep.
note the gradient is from `radius-halfStrokeWidth to radius+halfStrokeWidth`




![img.png](img.png)

```js
/* let g = [
    [0, 'red'],
    [0.5, 'blue'],
    [1, 'green']
  ]*/
  let g = [
    [0, '#00000000'],
    [0.4, 'rgba(238,0,255,0.27)'],
    [0.49, 'rgba(121,6,129,0.59)'],
    [0.5, 'rgba(0,0,0,0.48)'],
    [0.55, 'rgba(0,0,0,0.51)'],
    [0.56, 'rgba(75,234,229,0.56)'],
    [0.8, 'rgba(98,248,220,0.34)'],
    [0.95, 'rgba(98,248,220,0.12)'],
    [1, '#00000000'],
  ]

  let toonOptions = {
    height: 242,
    width: 242,
    // topojsonSrc: "../src/world.json",

    // globeBlur: true,
    water: "rgba(0,150,231,0.34)",
    globeStroke: g,
    // drawGlobeGradient: true,
    globeStrokeWidth: 25,

    land: "rgba(62,255,8,0.61)",
    landStroke: "#031703",
    // shadow: "rgba(129,35,129,0.63)",
    landStrokeWidth: 1,
  }
  var miniMap;

  function makeMiniMap(theme) {

    let o = Object.assign({
      height: 40,
      width: 40,
      topojsonSrc: "../src/world.json",
      onAdd: (map, mm) => {
        mm._container.addEventListener("click", e => {
          console.log("World CLicked");
          meSize += 5; // just a demo of detecting click .
          miniMap.hackMation(30, meSize)
        })
      }
    }, theme)

    miniMap = new L.Control.GlobeMiniMap(o).addTo(map);


  }
  let meSize = 40
  makeMiniMap(toonOptions);

```

## Attribution

I saw this static 3d minimap and thought it would be a great plugin for leaflet.  [Static 3d Minimap](http://earthview.withgoogle.com/marshall-islands-6155)

The globe is built with d3, and is based on [World Tour](http://bl.ocks.org/mbostock/4183330) an excellent block by Mike Bostock.

The code for the plugin was modified from [Leaflet.MiniMap](https://github.com/Norkart/Leaflet-MiniMap) by Robert Nordon

The marker SVG on the minimap is from [fontawesome via wikimedia.org](https://upload.wikimedia.org/wikipedia/commons/9/93/Map_marker_font_awesome.svg)

## Issues / todo

- Need a smooth resize animation.
- need rotation support
- need a way of mapping the pointer location on the globe to a latlon.


Please report issues on [github](https://github.com/chriswhong/leaflet-globe-minimap/issues).
I am a noob at this plugin business, and am not sure if the /dist and build stuff is working properly, so advice and Pull Requests are welcome!


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).
## 0.0.3 broken fixed in 0.0.4
### Added
  - updated pages / readme
  - fix bug with resize
  - add some option validation
  - `options.radius` is in testing now overrides width and height if used

## 0.0.2
### Added
- `topojsonSrc` option to specify custom TopoJSON world-map data
- `setSize(w,h)` update options and re add to map to rerender and compute
- `hackMation()` try it for a demo hacky animation of resize todo figure out how to do this fast with css.
- `transitionMap(pos, duration)` added duration option.
- `options.onAdd / onRemove` calbacks with map, miniMap, for adding you events to the dom.

### Changed
- Also made things correctly scale, note width and height must be = for now due to d3 mathyness. todo fix :)
- Fixed bug with size change and transparency (issue)[https://github.com/chriswhong/leaflet-globeminimap/issues/12]
- Upgrade dependencies: d3, leaflet, topojson(-client), uglify-js
- Example page loads dependencies from `node_modules/` instead of making network requests
- Convert to ES6+ module
- Convert to ES6+ class
- Use Webpack for build process (instead of directly calling uglify-js)
- Example page loads plugin from `dist/` instead of `src/`

## 0.0.1 - 2015-10-10
### Added
- Thanks for the es6 update https://github.com/stilist/leaflet-globeminimap
- Forked from https://github.com/chriswhong/leaflet-globeminimap
- Forked from https://github.com/Norkart/Leaflet-MiniMap


src/Control.GlobeMiniMap.js
import d3 from 'd3';
import L from 'leaflet';
import 'topojson-client';


class MiniMap extends L.Control {
  //layer is the map layer to be shown in the minimap
  initialize(options) {
    L.Util.setOptions(this, options);

    console.assert(typeof this.options.width === 'number', "width must be number px")
    console.assert(typeof this.options.height === 'number', "height must be number px")
    console.assert(typeof this.options.duration === 'number', "duration must be number ms")

    if (this.options.height !== this.options.width) {
      console.warn("You have asked for a different height then width, this is currently buggy only circles.")
    }
    if (typeof this.options.radius === "number") {
      this.options.width = this.options.radius * 2;
      this.options.height = this.options.radius * 2;
    }

  }

  onAdd(map) {
    // console.log('onAdd()');

    this._mainMap = map;

    //Creating the container and stopping events from spilling through to the main map.
    this._container = L.DomUtil.create('div', 'leaflet-control-minimap');

    //if width is specifed we need to pad by this;
    this._pad = this.options.globeStrokeWidth || 0
    if (this.options.globeStrokeWidth === true) {
      this._pad = this.options.width / 5;
    }
    //we need more padding if the blur is turned on.
    if (this.options.globeBlur) this._pad *= 2;

    this._container.style.width = this.options.width + this._pad + 'px';
    this._container.style.height = this.options.height + this._pad + 'px';


    L.DomEvent.disableClickPropagation(this._container);
    L.DomEvent.on(this._container, 'mousewheel', L.DomEvent.stopPropagation);

    //Keep a record of this to prevent auto toggling when the user explicitly doesn't want it.
    this._userToggledDisplay = false;
    this._minimized = false;

    this._mainMap.on('moveend', this._onMainMapMoved, this);

    return this._container;
  }

  addTo(map) {
    // console.log('addTo()');
    L.Control.prototype.addTo.call(this, map);
    this.initCanvas();

    if (typeof this.options.onAdd === "function") {
      this.options.onAdd(map, this);
    }
    return this;
  }

  initCanvas() {
    //marker icon
    //https://upload.wikimedia.org/wikipedia/commons/9/93/Map_marker_font_awesome.svg
    let w = this.options.width;
    let h = this.options.height;
    let pw = w + this._pad;
    let ph = w + this._pad;
    let p = this._pad;
    let p2 = p / 2;

    let sx = .01 * pw / 82
    let sy = .01 * ph / 82
    let mw = 400
    let ox = (pw * 0.5) / sx;
    let oy = (ph * 0.5) / sy
    let offset = `scale(${sx},-${sy}),translate(${ox - (mw)},-${oy})`;
    console.log("offset: ", offset)
    d3.select('.leaflet-control-minimap')
      .append('svg')
      .attr("width", pw)
      .attr("height", ph)
      .attr("style", "position: absolute; left: 0; top: 0;")
      .append('path')
      .attr('d', 'm 768,896 q 0,106 -75,181 -75,75 -181,75 -106,0 -181,-75 -75,-75 -75,-181 0,-106 75,-181 75,-75 181,-75 106,0 181,75 75,75 75,181 z m 256,0 q 0,-109 -33,-179 L 627,-57 q -16,-33 -47.5,-52 -31.5,-19 -67.5,-19 -36,0 -67.5,19 Q 413,-90 398,-57 L 33,717 Q 0,787 0,896 q 0,212 150,362 150,150 362,150 212,0 362,-150 150,-150 150,-362 z')
      .attr('transform', offset)
      .attr('style', 'fill:' + this.options.marker);


    let s = (w + h) / 4; // no point broken with difrent size
    //todo allow indapendant scaling so difrent numbers doesnt break it
    //https://github.com/d3/d3-zoom/issues/48

    this.projection = d3.geo.orthographic()
      .scale(s)

      // .transform()
      // .scaleLock([w, h])
      .translate([(pw) / 2, (ph) / 2])
      .rotate([0, 0])
      .clipAngle(90);

    var canvas = d3.select('.leaflet-control-minimap').append("canvas")
      .attr("width", pw)
      .attr("height", ph)

    this.c = canvas.node().getContext("2d");

    this.path = d3.geo.path()
      .projection(this.projection)
      .context(this.c);

    d3.json(this.options.topojsonSrc, (world) => {
      this.globe = {type: "Sphere"};
      this.land = topojson.feature(world, world.objects.land);
    });

    //set to current view
    this.transitionMap(this._mainMap.getCenter(), 0);
  }

  hackMation(s, e, self) {
    self = self || this
    s = s || 40
    e = e || 400
    if (s < e) {
      s += 3
      self.setSize(s);
      setTimeout(self.hackMation, 15, s, e, self);
    }

  }

  /**
   * Animate the map to a position over d ms
   * @param point - L.LatLng
   * @param duration - duration in ms
   */
  transitionMap(point, duration) {
    if (typeof duration != "number") {
      duration = duration || this.options.duration
    }
    // console.log('transtionMap');
    var that = this;
    var c = that.c;
    var path = that.path;


    function drawGlobe() {
      //custom predraw background
    }

    function drawLand() {
      //cool

    }


    d3.transition()
      .duration(duration)
      .each("start", function () {
      })
      .tween("rotate", function () {
        let r = d3.interpolate(that.projection.rotate(), [-point.lng, -point.lat]);
        let pw = that.options.width + that._pad;
        let ph = that.options.height + that._pad;
        return function (t) { // draw function

          that.projection.rotate(r(t));

          //clear
          c.clearRect(0, 0, pw, ph);

          drawGlobe();
          drawLand();
          //draw base circle
          if (that.options.globeFilter) {
            c.filter = that.options.globeFilter
          }
          c.fillStyle = that.options.water;
          c.beginPath();
          path(that.globe);
          c.fill();

          //allow us to specify a outline of the world
          // suports globeStroke= color, globeFStrokeFilter takes prioraty
          // but there is a preset globeBlur
          // you can set globeStrokeWidth
          if (that.options.globeStroke) {
            c.save()

            let g = that.options.globeStrokeWidth
            if (g === true) { // handle dynamic
              g = that.options.width / 5;
            } else if (typeof g !== "number") {
              g = 2; // default if not defined
            }
            c.lineWidth = g;

            if (that.options.globeStrokeFilter) {
              c.filter = that.options.globeStrokeFilter;
            } else if (that.options.globeBlur) {
              var h = that.options.globeBlur;
              if (h === true) h = g / 4;
              c.filter = 'blur(' + h + 'px)'
            }


            let style = that.options.globeStroke;
            if (Array.isArray(style)) {
              //we will construct a radial gradient based on this
              let center = pw / 2
              let w2 = that.options.width / 2;
              let halfStrokeWidth = g / 2;
              let r0 = w2 - halfStrokeWidth;
              if (r0 < 0) {
                r0 = 0;
                console.warn("createRadialGradient Negative radius not allowed. setting to 0!");
              }
              let gradient = c.createRadialGradient(center, center, r0, center, center, w2 + halfStrokeWidth)
              style.forEach(s => {
                console.log("adding step: ", ...s);
                gradient.addColorStop(...s);
              })


              style = gradient;

              if (that.options.drawGlobeGradient) {
                console.log("debug draw")
                c.save()
                c.fillStyle = style;
                c.fillRect(0, 0, pw, ph);
                c.restore()
              }
            }


            c.strokeStyle = style

            c.beginPath();
            path(that.globe);
            c.stroke();
            c.restore();
          }

          //ok now onto drawing the land on top of the circle

          if (that.options.landFilter) {
            c.filter = that.options.landFilter
          } else if (that.options.landShadow) {
            let s = that.options.landShadow;
            let p = pw / 100;
            let l = pw / 100;
            c.filter = 'drop-shadow(' + l + 'px ' + l + 'px ' + p + 'px ' + s + ')'

          } else {
            c.filter = "";
          }
          c.fillStyle = that.options.land,
            c.beginPath(),
            path(that.land),
            c.fill();

          if (that.options.landStroke) {
            c.save();
            if (that.options.landStrokeFilter) {
              c.filter = that.options.landStrokeFilter
            } else {
              c.filter = "";
            }
            c.strokeStyle = that.options.landStroke
            c.lineWidth = that.options.landStrokeWidth || 2
            c.beginPath();
            path(that.land);
            c.stroke()
            c.restore();
          }
        };
      })
  }

  onRemove(map) {
    this._mainMap.off('moveend', this._onMainMapMoved, this);
    this._mainMap.off('move', this._onMainMapMoving, this);

    if (typeof this.options.onRemove === "function") {
      this.options.onRemove(map, this);
    }
  }

  //note size must be square for now.
  setSize(w, h) {
    this.options.width = w;
    this.options.height = h || w;
    if (this._map) {
      //force recompute hack
      this.remove().addTo(this._mainMap);
    }

  }

  _onMainMapMoved(e) {
    // console.log('mainmapmoved');
    if (!this._miniMapMoving) {
      this._mainMapMoving = true;

      this.transitionMap(this._mainMap.getCenter());
    } else {
      this._miniMapMoving = false;
    }
  }

}

MiniMap.prototype.options = {
  position: 'bottomright',
  width: 82,
  height: 82,
  land: "#bbb",
  water: "rgba(0, 0, 0, 0.3)",
  marker: "#CC0000",
  topojsonSrc: 'data/world.json',
  duration: 1250,
  onAdd: false, //(map, this) for adding event listners.
  onRemove: false, // callback for cleaning up on remove
  globeStroke: false, // the color to set for the strokeStyle.
  // implementation note. this._pad is the padding added to the div.
  globeStrokeWidth: 4, // true for 1/5 width,
  globeBlur: false, // preset to pixel for gausian blur on the stroke
  globeStrokeFilter: false, // override stroke filter string.
  globeFilter: false,
  landStroke: false,
  landStrokeWidth: 1,
  landShadow: false, // override filter with a shadow of color
  landFilter: false, // idk try it. https://developer.mozilla.org/en-US/docs/Web/CSS/filter
  landStrokeFilter: false,


}

L.Control.GlobeMiniMap = MiniMap

export default MiniMap

L.control.globeminimap = function (layer, options) {
  return new L.Control.GlobeMiniMap(layer, options);
};

L.Map.mergeOptions({
  miniMapControl: false
});

L.Map.addInitHook(function () {
  if (this.options.miniMapControl) {
    this.miniMapControl = (new GlobeMiniMap()).addTo(this);
  }
});
